src/imputation/impute_civ_def.py

```diff
@@ -0,0 +1,248 @@
+
+    Returns:
+        Dict[str, Tuple(float, float)]
+    """
+    # create dictionary to hold civil or defence ratios for each class
+    pgsic_dict = {}
+
+    # Filter out imputation classes that are missing either "201" or "rusic"
+    # and exclude empty pg_sic classes
+    cond1 = ~(df["pg_sic_class"].str.contains("nan")) & (
+        df["empty_pgsic_group"] == False
+    )
+    filtered_df = df[cond1]
+
+    # Group by the pg_sic imputation class
+    pg_sic_grp = filtered_df.groupby("pg_sic_class")
+
+    # loop through the pg_sic imputation class groups
+    for pg_sic_class, class_group_df in pg_sic_grp:
+        pgsic_dict[pg_sic_class] = calc_cd_proportions(class_group_df)
+
+    # create a second dictionary to hold civil or defence ratios
+    # for the "empty_pgsic_group" cases
+    pg_dict = {}
+
+    # filter out invalid pg classes and empty pg groups from the original dataframe
+    cond2 = ~(df["pg_class"].str.contains("nan")) & (df["empty_pg_group"] == False)
+    filtered_df2 = df[cond2]
+
+    # evaluate which pg classes are needed for empty pg_sic groups
+    num_empty = filtered_df2.groupby("pg_class")["empty_pgsic_group"].transform(sum)
+    filtered_df2 = filtered_df2.loc[num_empty > 0]
+
+    # Group by the pg-only imputation class the loop through the groups
+    pg_grp = filtered_df2.groupby("pg_class")
+
+    for pg_class, class_group_df in pg_grp:
+        pg_dict[pg_class] = calc_cd_proportions(class_group_df)
+
+    return pgsic_dict, pg_dict
+
+
+def calc_empty_group(
+    df: pd.DataFrame, class_name: str, new_col_name: str
+) -> pd.DataFrame:
+    """Add a new bool column flagging empty groups."""
+
+    clear_mask = df["status"].isin(clear_statuses)
+    df["valid_civdef_val"] = ~df["200"].isnull() & clear_mask
+
+    # calculate the number of valid entries in the class for column 200
+    num_valid = df.groupby(class_name)["valid_civdef_val"].transform(sum)
+
+    # exclude classes that are not valid
+    valid_class_mask = ~(df[class_name].str.contains("nan"))
+
+    # create new bool column to flag empty classes
+    df[new_col_name] = valid_class_mask & (num_valid == 0)
+
+    df = df.drop("valid_civdef_val", axis=1)
+
+    return df
+
+
+def prep_cd_imp_classes(df: pd.DataFrame) -> pd.DataFrame:
+    """Create new columns to use for imputation classes,
+
+    Also create boolean columns to flag for when these contain no
+    valid entries that can be used for imputation.
+    """
+    # create imputation classes based on product group and rusic
+    df = tmi.create_imp_class_col(df, "201", "rusic", "pg_sic_class")
+
+    # flag empty pg_sic_classes in new bool col "empty_pgsic_class"
+    df = calc_empty_group(df, "pg_sic_class", "empty_pgsic_group")
+
+    # create a new imputation class called "pg_class" based on product group
+    # (col 201) only, to be used when imputation cannot be performed using
+    # both prodcut group and SIC.
+    df = tmi.create_imp_class_col(df, "201", None, "pg_class")
```

Comment:
Wondering about using 201 or df["pg_class"] = df["201]

Reply:
That nearly works but not quite, as it won't cover the cellno 817 special case.